File: VB/MessageBoxHelper/MyMessageBox/MyXtraMessageBox.vb
Imports DevExpress.XtraEditors
Imports MessageBoxHelper.Internal
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Reflection
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace MessageBoxHelper
    Public NotInheritable Class MyXtraMessageBox

        Private Sub New()
        End Sub

        Public Shared Function Show(ByVal options As MyXtraMessageArgs) As DialogResult
            Dim messageForm As New MyXtraMessageBoxForm()
            Dim result As DialogResult = messageForm.ShowForm(options)
            options.ShowMessageNextTime = messageForm.ShowMessageNextTime
            Return result
        End Function

        Public Shared Function GetDialogResultsFromButtons(ByVal buttons As MessageBoxButtons) As DialogResult()
            Dim xtraMessageBoxInfo As MethodInfo = GetType(XtraMessageBox).GetMethod("MessageBoxButtonsToDialogResults", BindingFlags.Static Or BindingFlags.NonPublic Or BindingFlags.Public)
            Return DirectCast(xtraMessageBoxInfo.Invoke(Nothing, New Object() { buttons }), DialogResult())
        End Function

        Public Shared Function GetIcon(ByVal iconInfo As String) As Icon
            Dim icon As Object = System.Enum.Parse(GetType(MessageBoxIcon), iconInfo)
            Dim xtraMessageBoxInfo As MethodInfo = GetType(XtraMessageBox).GetMethod("MessageBoxIconToIcon", BindingFlags.Static Or BindingFlags.NonPublic Or BindingFlags.Public)
            icon = xtraMessageBoxInfo.Invoke(Nothing, New Object() { icon })
            Return DirectCast(icon, Icon)
        End Function
    End Class
End Namespace

File: CS/MessageBoxHelper/MyMessageBox/MyXtraMessageBox.cs


using DevExpress.XtraEditors;
using MessageBoxHelper.Internal;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace MessageBoxHelper {
    public static class MyXtraMessageBox {
        public static DialogResult Show(MyXtraMessageArgs options) {
            MyXtraMessageBoxForm messageForm = new MyXtraMessageBoxForm();
            DialogResult result = messageForm.ShowForm(options);
            options.ShowMessageNextTime = messageForm.ShowMessageNextTime;
            return result;
        }

        public static DialogResult[] GetDialogResultsFromButtons(MessageBoxButtons buttons) {
            MethodInfo xtraMessageBoxInfo = typeof(XtraMessageBox).GetMethod("MessageBoxButtonsToDialogResults", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
            return (DialogResult[])xtraMessageBoxInfo.Invoke(null, new object[] { buttons });
        }

        public static Icon GetIcon(string iconInfo) {
            object icon = Enum.Parse(typeof(MessageBoxIcon), iconInfo);
            MethodInfo xtraMessageBoxInfo = typeof(XtraMessageBox).GetMethod("MessageBoxIconToIcon", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
            icon = xtraMessageBoxInfo.Invoke(null, new object[] { icon });
            return (Icon)icon;
        }
    }
}
